### DISCUSSION TITLE
Initial test scenarios

### DISCUSSION THREAD

**Discussion by rachel-fenichel** [RaspberryPiFoundation/blockly-keyboard-experimentation#57] [General]
The goal in the first phase of this project is to make it possible to complete introductory curriculum sequences using only keyboard navigation.

Each educational product has a different introductory sequence, but most involve creating between one and four stacks of blocks, with each stack containing between one and five blocks. (For example, here are the [MakeCode + micro:bit introductory tutorials](https://makecode.microbit.org/tutorials).) It is normal to teach sequencing, loops, and sometimes conditionals in these early stages. Usually there is a clear visual or aural output that a student can use to check their work.

For testing purposes we need to define a simple sequence that can be completed on the [keyboard-experimentation test page](https://google.github.io/blockly-keyboard-experimentation/). I propose this because it lets us avoid issues of integration (e.g. linking beta versions of multiple libraries together in order to deploy a beta of MakeCode that depends on a rapidly changing keyboard-experimentation plugin).

The test page has a canvas on the left side and a Blockly workspace on the right side, with a limited number of blocks in a simple (non-category) flyout.
![image](https://github.com/user-attachments/assets/1ba1b795-32fc-4aad-9480-544b591a7323)

### Proposal

I propose the following introductory tasks, which may need new blocks. 

Note that these examples use visual outputs. That's entirely based on what I know how to do with p5.js this week, and is not a hard requirement. I consider these tests relevant because they can equally apply to a student with a broken mouse or bad touchpad. 

#### User tasks

1. Open and run the starter code, which will draw a blue background and a red circle.
![A Blockly workspace with blocks to create a blue background and draw a red circle](https://github.com/user-attachments/assets/25f65397-a0a7-4b27-b265-1122c0a9ad48)

2. Change the color of the background and the color of the circle, run the code, and verify the result.
![A Blockly workspace with blocks to create an orange background and draw a black circle](https://github.com/user-attachments/assets/039203e5-b768-456a-9558-df4948cd0b51)

3. Replace the color of the circle with a `random colour` block from the flyout.
![A Blockly workspace with blocks to create an orange background and draw a circle with a random colour](https://github.com/user-attachments/assets/2cd7a252-73e7-4b84-8636-17335bbb8cf8)

4. Modify the code to draw a heart emoji in addition to the circle.
![A Blockly workspace with blocks to create an orange background, draw a circle with a random colour, and draw a heart emoji](https://github.com/user-attachments/assets/a788bb3e-7b8e-473b-972b-5a44c913c187)

5. Modify the code to draw a bear emoji instead of a heart emoji
![A Blockly workspace with blocks to create an orange background, draw a circle with a random colour, and draw a bear emoji](https://github.com/user-attachments/assets/e670c21b-a507-412c-ab61-3fd9d613eee9)

6. Modify the code to only draw a bear emoji, with no circle
![A Blockly workspace with blocks to create an orange background and draw a bear emoji](https://github.com/user-attachments/assets/4378b907-47ff-400e-b393-359181fa1516)

#### Required capabilities

This sequence requires the following abilities. Where a discussion already exists, I've linked it to the appropriate action. We can create new discussions as needed.

- [Open the page and run code](https://github.com/google/blockly-keyboard-experimentation/discussions/49)
- [Navigate between stacks](https://github.com/google/blockly-keyboard-experimentation/discussions/17)
- [Navigate within a stack to a target block](https://github.com/google/blockly-keyboard-experimentation/discussions/20) 
- Open a complex field editor (colour picker) and enter a new value
- Open a moderately complex field editor (dropdown) and enter a new value
- [Navigate between blocks in the toolbox](https://github.com/google/blockly-keyboard-experimentation/discussions/18)
- Add a new block to the workspace and [move it to a target location in a stack](https://github.com/google/blockly-keyboard-experimentation/discussions/55)
  - Note that this can be done either by choosing a block first or by choosing a target location first. Both should be investigated.
- Add a new block to the workspace and [move it to replace a shadow block](https://github.com/google/blockly-keyboard-experimentation/discussions/55)
  - Note that this can be done either by choosing a block first or by choosing a target location first. Both should be investigated.
- Delete a block

### Future work

Obviously the list above is not exhaustive. handling text inputs and creating new top-level stacks are clear requirements that can follow from this initial work. Conditionals and loops (c-shape and e-shape blocks) are also often introduced at an early stage.

**Comment by microbit-lucy** [comment]
This looks great. I would be really interested to see those blocks in the experiment so we can walk through the test scenario end-to-end. I have shared it with our education team in case they have other thoughts to share but we will start to put together a test plan for user testing at the end of October/beginning of November based on this. Thanks.

**Reply by rachel-fenichel** [reply]
https://github.com/google/blockly-keyboard-experimentation/pull/58 tracks updating the page to be accessible to you.
